### PR TITLE
Generic git url support issue152

### LINKED ISSUE #152: Generic Git URL support
Hello @ericmdantas,
         As currently the git url references are hard coded to github, I was thinking of a feature where the user can enter a git url of their choice like BitBucket or a private git repo. This would essentially replace/add to the username prompt. Please let me know what you think and I can help on this.

Regards,
Rahul Swaminathan

### PR DESCRIPTION
**Pull request by rahulswaminathan**
Hi @ericmdantas ,
    This is an initial pull request which attempts to resolve #152 . I've some doubts regarding the changes to be made in tests _test/acceptance/test-app.js_ and _test/unit/test-generator.js_. This is my first pull request on GitHub :) so I might have made quite a few mistakes. Please let me know as I would like to correct them and learn.

Regards,
Rahul Swaminathan

### CODE REVIEW

**Comment by ericmdantas** [comment]
Hello, @rahulswaminathan!

It's looking good, man. Thanks for the hard work! 

I'll be able to take a closer look when I get home, then I'll be able to help you out with both unit and acceptance tests.

File: app/templates/_bower.json [hunk 1]
```diff
@@ -2,8 +2,8 @@
   "name": "<%= app %>",
   "main": "<%= app %>.js",
   "version": "0.0.1",
-  "homepage": "https://github.com/<% username %>/<%= app %>",
-  "authors": ["<%= username%> <email>"],
+  "homepage": "https://<% repohosturl %>/<% usernamespace %>/<%= app %>",
```
Comment (ericmdantas) [inline app/templates/_bower.json]: Missing `=` before `repohosturl` and `usernamespace` (the latter was my fault, :laughing:)

File: _ng/full/generator.js [hunk 2]
```diff
@@ -95,30 +111,73 @@ exports.MainGenerator = class MainGenerator {
...
+            name: 'userName',
+            message: 'What is your username?',
+            default: 'user.name'
           },
           {
-            name: 'githubUsername',
-            message: 'What is your username on Github?',
-            default: 'some-username-here'
+            name: 'userEmail',
+            message: 'What is your email-id?',
```
Comment (ericmdantas) [inline _ng/full/generator.js]: `What is your email?` sounds better.

**Comment by ericmdantas** [comment]
All in all, things are looking solid, just a few points:

- The tests seem to be failing because in some places the property is being created `camel cased` and being used as a var inside a file `lower cased`. Such as `{somethingHere: "a"}` and `<%=somethinghere%>`. Where the correct would be `<%=somethingHere%>`;
- I suggest we use username being lower cased - seems to be the most used;
- And then maybe we'd use `usernameNamespace` being camel cased? For the gitlab local service.
- `repoHostUrl`, `userEmail` and `usernameNamespace` should be case sensitive - in the generator properties and in the template files;

Let me know if you need anything.

**Comment by ericmdantas** [comment]
Now the tests are failing because there's no `repoHostUrl` in `test/acceptance/test-app.js`, in the `withPrompt` function ([here, for example](https://github.com/ericmdantas/generator-ng-fullstack/blob/master/test/acceptance/test-app.js#L122)):

```js
.withPrompts({
              appName: "a",
              githubUsername: "b",
              server: "node",
              transpilerServer: 'node',
              stack: 'fullstack',
              client: 'ng1'
            })
```

This is where you "simulate" what the user is trying to choose in the generator, so, you could:

```js
.withPrompts({
              appName: "a",
              githubUsername: "b",
              server: "node",
              transpilerServer: 'node',
              stack: 'fullstack',
              repoHostUrl: "github", // here
              client: 'ng1'
            })
```

And that's where you'd also mention the gitlab repo and others.

**Comment by ericmdantas** [comment]
You can run the tests locally, by the way.

```shell
$ npm t
```

**Comment by rahulswaminathan** [comment]
Thanks @ericmdantas ! This is where my doubt was. I'll test it locally and then commit it.

Regards,
RS

**Comment by ericmdantas** [comment]
Another useful thing I forgot to mention is:

If you want to play with your commits (trying out your changes to the code), you can run `npm link` in the root of the application and then every change you make will work when you run `yo ng-fullstack` and so on.

**Comment by ericmdantas** [comment]
Could you add some tests for bitbucket/gitlab? In `test/acceptance/test-app.js`, that is.

**Comment by ericmdantas** [comment]
I was trying it locally and it seems to be working just fine. I'll merge it manually.

Once again, thanks for the PR.

**Comment by ericmdantas** [comment]
Merged in bd888b77ad5d7d70826b0ad93ee73e48000095af.

**Comment by rahulswaminathan** [comment]
Sorry @ericmdantas , I got busy at work and couldn't add the tests. Thanks for your help!

**Pull request closed without merging**